apps/price_pusher/src/evm/evm.ts

```diff
@@ -223,42 +250,121 @@ export class EvmPricePusher implements IPricePusher {
-        gasPrice,
-        nonce: txNonce,
-        gasLimit: this.gasLimit,
-      })
-      .on("transactionHash", (hash: string) => {
-        this.logger.info({ hash }, "Price update successful");
-      })
-      .on("error", (err: Error, receipt?: TransactionReceipt) => {
-        if (err.message.includes("revert")) {
-          // Since we are using custom error structs on solidity the rejection
-          // doesn't return any information why the call has reverted. Assuming that
-          // the update data is valid there is no possible rejection cause other than
-          // the target chain price being already updated.
+    const pubTimesToPushParam = pubTimesToPush.map((pubTime) =>
+      BigInt(pubTime)
+    );
+
+    try {
+      const { request } =
+        await this.pythContract.simulate.updatePriceFeedsIfNecessary(
+          [priceFeedUpdateData, priceIdsWith0x, pubTimesToPushParam],
+          {
+            value: updateFee,
+            gasPrice: BigInt(Math.round(gasPrice)),
+            nonce: txNonce,
+            gas:
+              this.gasLimit !== undefined
+                ? BigInt(Math.round(this.gasLimit))
+                : undefined,
+          }
+        );
+
+      this.logger.debug({ request }, "Simulated request successfully");
+
+      const hash = await this.pythContract.write.updatePriceFeedsIfNecessary(
+        [priceFeedUpdateData, priceIdsWith0x, pubTimesToPushParam],
+        {
+          value: updateFee,
+          gasPrice: BigInt(Math.round(gasPrice)),
+          nonce: txNonce,
+          gas:
+            this.gasLimit !== undefined
+              ? BigInt(Math.round(this.gasLimit))
+              : undefined,
+          chain: this.walletClient.chain,
+          account: this.walletClient.account!,
+        }
+      );
+
+      this.logger.info({ hash }, "Price update sent");
+
+      this.waitForTransactionReceipt(hash);
+    } catch (err: any) {
+      this.logger.debug({ err }, "Simulating or sending transactions failed.");
+
+      if (err instanceof BaseError) {
+        if (
+          err.walk(
+            (e) =>
+              e instanceof ContractFunctionRevertedError &&
+              e.data?.errorName === "NoFreshUpdate"
+          )
+        ) {
           this.logger.info(
-            { err, receipt },
-            "Execution reverted. With high probability, the target chain price " +
-              "has already updated, Skipping this push."
+            "Simulation reverted because none of the updates are fresh. This is an expected behaviour to save gas. Skipping this push."
+          );
+          return;
+        }
+
+        if (err.walk((e) => e instanceof InsufficientFundsError)) {
+          this.logger.error(
+            { err },
+            "Wallet doesn't have enough balance. In a rare case, there might be issues with gas price " +
+              "calculation in the RPC."
+          );
+          throw new Error("Please top up the wallet");
```

Comment:
Minor nit but I suggest adding new error classes for something like this, it makes it much easier for consumers to disambiguate errors:

```ts
class PleaseTopUpWalletError extends Error {
  override message = "Please top up the wallet";
};
```

Although in this case I'd probably just re-throw the `InsufficientFundsError` myself...